### ISSUE TITLE
Migrating to v3

### ISSUE THREAD

**Issue by timfish** [getsentry/sentry-electron#370]
v3 of the Electron SDK includes significant changes to simplify usage and improve maintainability and bundler support. Migration should still be relatively painless!

- If you have trouble migrating, please respond in this issue
- If you find bugs with v3 or you can see areas where it can be improved, please open a new issue

For some advanced technical detail, you may want to check out the relevant [Proposal](https://github.com/getsentry/sentry-electron/issues/360) and [Pull Request](https://github.com/getsentry/sentry-electron/pull/361).

New Features:
- Session tracking now enabled by default
- Preload script no longer required [for most scenarios](https://github.com/getsentry/sentry-electron/issues/376)
- Optional relative imports for main/renderer/preload entry points
- Offline support
- Additional device context
- Minidumps for GPU crashes
- Browser Tracing now compatible

Major breaking changes:
- Native crashes now consider `sampleRate` and `beforeSend`
- Configuration through integrations rather than options

## Session Tracking

Session tracking is now enabled by default so you will see [Release Health](https://docs.sentry.io/product/releases/health/) data. Session tracking is via the `MainProcessSession` integration which as the name suggests, tracks sessions as per the main process lifetime. 

```ts
Sentry.init({
  dsn: '__DSN__',
  release: '__RELEASE__'
});
```

If you don't want to track sessions, this can be disabled by setting `autoSessionTracking` to `false`.

```ts
Sentry.init({
  dsn: '__DSN__',
  autoSessionTracking: false
});
```

If you have a use case where sessions should be tracked in a different way, please open an issue!

## Relative Imports

The SDK uses [multiple `package.json` fields](https://github.com/timfish/sentry-electron/blob/5bba1e9221e86f9d43f616f8db8112a23f22aea5/package.json#L5-L25) to ensure that bundlers automatically pickup the the correct entry point for each Electron process when using the root import (`const Sentry = require('@sentry/electron')` or `import * as Sentry from '@sentry/electron'`). This allows you to have a `sentry.js` with your Sentry configuration imported into every process.

However, not all bundlers are created equal and you may want to add specific integrations to only one of the Electron processes.

To support more complex configurations, you can now skip this automatic bundler target detection and import the process specific code directly:

### Main Process
```ts
const Sentry = require('@sentry/electron/main');
// or
import * as Sentry from '@sentry/electron/main';
```

### Renderer Process
```ts
const Sentry = require('@sentry/electron/renderer');
// or
import * as Sentry from '@sentry/electron/renderer';
```

### Preload Code
```ts
require('@sentry/electron/preload');
// or
import '@sentry/electron/preload';
```

## Offline Support
The `ElectronOfflineNetTransport` is now the default transport. It wraps `ElectronNetTransport` and saves payloads to disk if they cannot be sent

## Additional device context
The default enabled `AdditionalContext` integration includes additional device context like screen resolution and memory usage.

## Browser Tracing

The significant refactor now allows the use of Sentry browser tracing in the renderer process:

`main.js`
```ts
import * as Sentry from '@sentry/electron/main';

Sentry.init({
  dsn: '__DSN__'
}); 
```

`renderer.js`
```ts
import * as Sentry from '@sentry/electron/renderer';
import { Integrations } from '@sentry/tracing';

Sentry.init({
  integrations: [new Integrations.BrowserTracing()],
  tracesSampleRate: 1.0,
}); 
```

## Preload Scripts

**As of 3.0.0-beta.2 a preload script is no longer required for Electron >= v5**

## Native Crashes, `sampleRate` and `beforeSend`

Previously, the SDK did not consider `sampleRate` when sending native crash events and it was not possible to intercept them via the `beforeSend` hook. Theses are now correctly handled.

## Integrations over Options

Previously, the Electron SDK had various configuration options. Most of these have moved to integrations.

With v3, the only Electron specific configuration options are in the main process and are detailed below:
```ts
export interface ElectronMainOptions extends NodeOptions {
  /**
   * Inter-process communication mode to receive event and scope from renderers
   *
   * IPCMode.Classic - Configures Electron IPC
   * IPCMode.Protocol - Configures a custom protocol
   * IPCMode.Both - Configures both IPC and custom protocol
   *
   * Defaults to IPCMode.Both for maximum compatibility
   */
  ipcMode: IPCMode;
  /**
   * A function that returns an array of Electron session objects
   *
   * These sessions are used to configure communication between the Electron
   * main and renderer processes.
   *
   * Defaults to () => [session.defaultSession]
   */
  getSessions: () => Session[];
  /**
   * Callback to allow custom naming of renderer processes.
   *
   * If the callback is not set, or it returns `undefined`, the default naming
   * scheme is used.
   */
  getRendererName?: (contents: WebContents) => string | undefined;
}
```

#### Native Crash Reporting
For native crash reporting, you have three options
1. `SentryMinidump` integration (default)
Uploads minidump files via the Sentry Envelope endpoint with full breadcrumbs and context
2. `ElectronMinidump` integration 
Uploads minidumps via Crashpad/Breakpad built in uploader with partial context
```ts
import { init, Integrations} from '@sentry/electron';

init({
  dsn: '__DSN__',
  // Adding the ElectronMinidump integration like this 
  // ensures that it is the first integrations to be initialized
  integrations: defaultIntegrations => {
    return [new Integrations.ElectronMinidump(), ...defaultIntegrations]
  }
}); 
```
3.  No native crash reporting (remove the `SentryMinidump` integration)
```ts
import { init, Integrations } from '@sentry/electron';

init({
  dsn: '__DSN__',
  integrations: defaultIntegrations => {
    return ...defaultIntegrations.filter(i => i.name != Integrations.SentryMinidump.Id);
  }
}); 
```

#### `WebContents` `'unresponsive'` Events

By default, the `ElectronEvents` integrations captures events for unresponsive Electron renderers. If you would like to disable this:
```ts
import { init, Integrations } from '@sentry/electron';

init({
  dsn: '__DSN__',
  integrations: [new Integrations.ElectronEvents({ unresponsive: false }];
}); 
```

**Comment by varunsrin** [comment]
@timfish one issue when running it locally (but not in production) is: 

```
Fetch API cannot load sentry-ipc://sentry-electron.scope/sentry_key. URL scheme must be "http" or "https" for CORS request.
```

Another issue that would benefit from documentation is that there is a new header being sent, which may need to be configured if you have strict CORS settings: 

```
Access to XMLHttpRequest at 'http://localhost:3000/...' from origin 'http://...' has been blocked by CORS policy: Request header field sentry-trace is not allowed by Access-Control-Allow-Headers in preflight response.
```

**Comment by timfish** [comment]
Thanks for letting me know!

We register the protocol [to bypass CSP](https://github.com/getsentry/sentry-electron/blob/e340110ac4f23cefa2221a397bb8f07459c86657/src/main/ipc.ts#L73) but it looks like I also need to add `corsEnabled` too. None of the [`privileges`](https://www.electronjs.org/docs/latest/api/structures/custom-scheme) are documented but I'm guessing this will remove the CORS issues.

**Comment by timfish** [comment]
@varunsrin The CORS fix has been shipped in 3.0.0-beta.3. Let me know if it fixes those warnings for you!

**Comment by varunsrin** [comment]
just tested, and it still throws the same warnings.

**Comment by timfish** [comment]
@varunsrin To help me better reproduce this issue and add a test case, can you give more details of your setup?

I'm guessing during development you're loading from localhost? 

> which may need to be configured if you have strict CORS settings

What kind of strict CORS settings are you referring to?

**Comment by Kyusung4698** [comment]
@timfish Hi, thanks for putting in the work for v3. I just got a quick question about this statement:
> As of 3.0.0-beta.2 a preload script is no longer required for Electron >= v5

Do we still need to include it via `import '@sentry/electron/preload';` or not? Regardless of what I do, I always get this error:
```
[14000:1203/144806.122:ERROR:CONSOLE(93)] "Unable to load preload script: ...\node_modules\@sentry\electron\preload\index.js", source: node:electron/js2c/sandbox_bundle (93)
[14000:1203/144806.122:ERROR:CONSOLE(93)] "Error: module not found: ../common/ipc", source: node:electron/js2c/sandbox_bundle (93)
```

It doesn't seem to affect the functionality tho...

**Comment by timfish** [comment]
@Kyusung4698 I've [opened a PR](https://github.com/getsentry/sentry-electron/pull/396) to fix this issue and error if it the build gets broken in the future.

Because we fall back to custom protocol this only resulted in logged errors rather than breaking functionality.

**Comment by Kyusung4698** [comment]
> @Kyusung4698 I've [opened a PR](https://github.com/getsentry/sentry-electron/pull/396) to fix this issue and error if it the build gets broken in the future.

Thanks! Could you please release a new package version once your PR gets approved? Also just to be sure, do we need to include the preload or not?

**Comment by timfish** [comment]
@Kyusung4698 I've not managed to do another release yet, but you can disable IPC which is triggering the warning and only use the custom protocol via:

```ts
const { init, IPCMode } = require('@sentry/electron');

init({
  dsn: '__DSN__',
  ipcMode: IPCMode.Protocol
});
```

Custom protocol works in Electron >= v5 and means you don't need to use preload.

**Comment by artplan1** [comment]
hi, I'm having this issue in 3.0.0-beta.3

![image](https://user-images.githubusercontent.com/9060346/147660478-9cfed851-1bdc-4cc3-8fdf-886c3b7e22f6.png)

![image](https://user-images.githubusercontent.com/9060346/147660664-ce150b01-d231-4ae7-bb5a-4c3f8db99855.png)

![image](https://user-images.githubusercontent.com/9060346/147660720-7000ae20-4fa7-4707-a8bf-d925283a4860.png)

changing `ipcMode` doesn't help

in main:

```js
import { init, IPCMode } from "@sentry/electron/main";

init({
  dsn: __DSN__,
  ipcMode: IPCMode.Protocol,
});
```

renderer:

```js
import { init } from "@sentry/electron/renderer";

init({
  debug: true,
});

setTimeout(() => {
  throw new Error("Some renderer error");
}, 500);
```

**Comment by timfish** [comment]
The `event processor returned null` logged error is expected but unfortunate. 

In the renderer, we essentially use `@sentry/browser` with an integration that intercepts events and sends them to the main process where additional context is added before they're sent to Sentry. That integration returns null to ensure that no further processing or sending occurs from the renderer. `@sentry/core` logs the above error to aid users debugging their integrations. 

In this case the event should still get sent to Sentry.

**Comment by artplan1** [comment]
got it, but in my case errors from renderer aren't sent to Sentry. what can I do to debug why it's happening?

**Comment by timfish** [comment]
Any interesting console output in the main process? `debug: true` in the main process should add some more info.

How does your app differ from the apps in `examples` and `test/e2e/test-apps`? These are the current integration tests. Could be a CSP or CORS error but would expect that to be logged in renderer.

Are you using anything more than the default session in renderers? We setup the protocol handler only for the default session be default. This can be overridden in the main options: https://github.com/getsentry/sentry-electron/blob/12058e46133f762bb7bdc873982727a3da813e88/src/main/sdk.ts#L52

**Comment by artplan1** [comment]
@timfish thank you for feedback, I figured that was causing events to not be sent - https://github.com/getsentry/sentry-electron/issues/403

events are sent with my fix, but can I mute `Sentry Logger [Error]: SentryError2: An event processor returned null, will not send event.` somehow?

**Comment by timfish** [comment]
Sentry logger should only display errors when `debug: true`.

**Comment by Kyusung4698** [comment]
Thanks for the new release! I'm getting quite a few 400 errors after upgrading to the 3.0.0-beta.4 release within the module itself. Is this known? There is not much information provided what the cause of the failed request is. Could you @timfish have a look? 

```
    "metadata": {
        "display_title_with_tree_label": false,
        "filename": "app:///node_modules/@sentry/electron/main/transports/electron-net.js",
        "function": "ClientRequest.<anonymous>",
        "type": "Error",
        "value": "HTTP Error (400)"
    },
    "sdk": {
        "name": "sentry.javascript.electron",
        "version": "3.0.0-beta.4",
        "integrations": [
            "ElectronBreadcrumbs",
            "ElectronMinidump",
            "CaptureConsole",
            "Net",
            "MainContext",
            "ChildProcess",
            "OnUncaughtException",
            "PreloadInjection",
            "AdditionalContext",
            "InboundFilters",
            "FunctionToString",
            "Console",
            "Http",
            "OnUnhandledRejection",
            "LinkedErrors",
            "MainProcessSession"
        ],
        "packages": [
            {
                "name": "npm:@sentry/electron",
                "version": "3.0.0-beta.4"
            }
        ]
    }
```

**Comment by timfish** [comment]
I would love to fix any potential issues with the latest SDK before it makes it to a stable release!

However, I have not personally observed the issues you're reporting and I haven't seen it with any of the e2e tests. 

I need more information or a repository that reproduces the issue!

Where is the JSON you've included coming from?

**Comment by Kyusung4698** [comment]
I took the JSON from an event in sentry and extracted the relevant parts. I wasn't able to reproduce it locally, only seeing those errors in our sentry issue list. 
![image](https://user-images.githubusercontent.com/38520678/149789348-29f3f1ac-ed3a-4583-b4f9-f067c199d44f.png)

**Comment by timfish** [comment]
Ah ok, so these events are making it to Sentry but they probably shouldn't be.

The latest beta includes offline support which means that events that cannot be sent are saved to disk and sent when a connection is detected. It's highly likely that your app was getting HTTP errors at this point in v2 of the SDK, but they would have never have made it to Sentry because the server was not accessible.

@AbhiPrasad how do you think I should go about improving this? Simply never queue errors originating from the transport?

Outside of SDK transport errors, the offline support is going to increase the number of errors users get due to intermittent/bad connections. Other than make this more clear in the docs, is there anything else we should to to improve this?

**Comment by Kyusung4698** [comment]
@timfish That's maybe a wild guess but we sometimes return null in `beforeSend` in order to not process spammy events. Could it be that if the network connection is unstable a 'null' event is stored on disk and then, once the connection is back, send as 'null' event resulting in the BadRequest response?

**Comment by timfish** [comment]
The offline transport (`ElectronOfflineNetTransport`) is simply a wrapper around the original transport (`ElectronNetTransport`) that queues events that fail to send. 

If `beforeSend` returns `null`, the event will be rejected by `@sentry/core` and is never passed to the transport for sending.

**Comment by AbhiPrasad** [comment]
> how do you think I should go about improving this? Simply never queue errors originating from the transport?

I can't ever see a case where these errors are of high value to users, I would just filter them out if possible. We do have the `SentryError` class for similar logic in the core JS SDK.

**Comment by chetbox** [comment]
Thanks for this handy guide!

From the above it seems like `ipcMode` and `getSessions` have default values but the Typescript types require that specified.

Example with `@sentry/electron` v2 we have:

```ts
  SentryElectron.init({
    dsn: ...,
    release: ...
  });
```

The same works with v3, but the Typescript types insist that `ipcMode` and `getSessions` are defined, even though they seem to have defaults in the implementation:

> Type '{ dsn: string; release: string; environment: "production" | "development"; ignoreErrors: string[]; normalizeDepth: number; }' is missing the following properties from type 'ElectronOptions': ipcMode, getSessionsts(2345)

Everything seems to work fine without `ipcMode` and `getSessions`.

**Comment by timfish** [comment]
Thanks for reporting this @chetbox, these should be optional!

**Comment by chetbox** [comment]
> Thanks for reporting this @chetbox, these should be optional!

Cheers for confirming so quick @timfish! There a PR here: https://github.com/getsentry/sentry-electron/pull/412

**Comment by timfish** [comment]
Many thanks for the PRs!

**Comment by timfish** [comment]
3.0.0 has now been released and this migration guide has moved to [here](https://github.com/getsentry/sentry-electron/blob/master/MIGRATION.md).

**Comment by acsant** [comment]
> Thanks for the detailed writeup Tim. I'm not migrating, but trying to integrate sentry for the first time and came across this thread. Unfortunately I wasn't successful at making it work.
> 
> I kept getting this error, and my preload script wouldn't build:
> 
> ```
> preload script must have absolute path: ./node_modules/@sentry/electron/esm/preload/index.js
> ```
> 
> I'm on Electron: 15.2.0 and using electron-forge 6.0.0-beta.61, including the webpack plugin. I dont think electron forge's webpack plugin is particularly unique, but as far as I understand, it bundles everything into the output files, so there's no `node_modules` at runtime--though that path might just be a module id so maybe that's a red herring.
> 
> fwiw I've tried:
> 
> * adding `@sentry/electron/preload/index.js` to the forge `prefixEntries` array, which basically just concats those files before the preload's source file as a webpack entry file.
> * adding a specific `webpack.resolve.alias` in the renderer webpack config to try and resolve sentry to an absolute path
> * adding `externals: { electron: 'commonjs electron' }` to the renderer webpack config (which I got from a different issue)
> 
> Happy to provide any other info that can help

Facing this issue as well
